File: frontend/lib/sentry.dart
import 'package:flutter/cupertino.dart';
import 'package:sentry_flutter/sentry_flutter.dart';

Future<void> runAppWithSentry(void Function() runApp) async {
  await SentryFlutter.init((options) {
    options.dsn = 'https://ceb1e25ecc334e26b1469a0bc325b7c9@sentry.tuerantuer.org/4';
    // https://docs.sentry.io/concepts/key-terms/tracing/
    options.tracesSampleRate = 0.05;
  }, appRunner: runApp);
}

Future<void> reportError(dynamic exception, dynamic stackTrace) async {
  _printError(exception, stackTrace);
  if (Sentry.isEnabled) {
    await Sentry.captureException(exception, stackTrace: stackTrace);
  }
}

void _printError(dynamic exception, dynamic stackTrace) async {
  if (stackTrace is StackTrace) {
    debugPrintStack(stackTrace: stackTrace, label: exception.toString());
  } else {
    debugPrint(exception.toString());
  }
}

File: backend/src/main/kotlin/app/ehrenamtskarte/backend/common/utils/Sentry.kt
package app.ehrenamtskarte.backend.common.utils

import io.opentelemetry.api.trace.propagation.W3CTraceContextPropagator
import io.opentelemetry.context.propagation.ContextPropagators
import io.opentelemetry.sdk.OpenTelemetrySdk
import io.opentelemetry.sdk.trace.SdkTracerProvider
import io.sentry.Sentry
import io.sentry.opentelemetry.OpenTelemetryLinkErrorEventProcessor
import io.sentry.opentelemetry.SentryPropagator
import io.sentry.opentelemetry.SentrySpanProcessor

fun initializeSentry() {
    OpenTelemetrySdk.builder()
        .setTracerProvider(
            SdkTracerProvider.builder()
                .addSpanProcessor(SentrySpanProcessor()) // Sendet direkt an Sentry
                .build(),
        )
        .setPropagators(
            ContextPropagators.create(
                io.opentelemetry.context.propagation.TextMapPropagator.composite(
                    W3CTraceContextPropagator.getInstance(),
                    SentryPropagator(),
                ),
            ),
        )
        .buildAndRegisterGlobal()

    Sentry.init { options ->
        options.dsn = "https://c0c177b8fa35e0fd90d788b0d02bcb2c@sentry.tuerantuer.org//7"
        // https://docs.sentry.io/concepts/key-terms/tracing/
        options.tracesSampleRate = 1.0
        // Adds tracing to errors
        options.addEventProcessor(OpenTelemetryLinkErrorEventProcessor())
        options.isDebug = true // Activate debug logs
    }
}

fun reportErrorToSentry(e: Exception) {
    if (Sentry.isEnabled()) {
        Sentry.captureException(e)
    }
}
